File: tcMenuJavaApi/src/main/java/com/thecoderscorner/menu/domain/util/MenuItemFormatter.java
package com.thecoderscorner.menu.domain.util;

import com.thecoderscorner.menu.domain.*;
import com.thecoderscorner.menu.domain.state.CurrentScrollPosition;
import com.thecoderscorner.menu.domain.state.PortableColor;
import com.thecoderscorner.menu.persist.LocaleMappingHandler;

import java.math.BigDecimal;

public class MenuItemFormatter {
    private static final MenuItemFormatter defaultInstance = new MenuItemFormatter();
    private LocaleMappingHandler localHandler = LocaleMappingHandler.NOOP_IMPLEMENTATION;

    public MenuItemFormatter(LocaleMappingHandler handler) {
        if(handler != null) localHandler = handler;
    }

    private MenuItemFormatter() {
    }

    public static MenuItemFormatter defaultInstance() {
        return defaultInstance;
    }

    public static void setDefaultLocalHandler(LocaleMappingHandler localHandler) {
        defaultInstance.localHandler = localHandler;
    }

    public String bundleIfPossible(String s) {
        return localHandler.getFromLocaleWithDefault(s, s);
    }

    public String formatToWire(MenuItem item, final String text) {
        return MenuItemHelper.visitWithResult(item, new AbstractMenuItemVisitor<String>() {
            @Override
            public void anyItem(MenuItem item) {
                throw new IllegalArgumentException(item + "not editable");
            }

            @Override
            public void visit(AnalogMenuItem item) {
                setResult(formatAnalogWire(item, text));
            }

            @Override
            public void visit(EnumMenuItem item) {
                setResult(formatEnumWire(item, text));
            }

            @Override
            public void visit(BooleanMenuItem item) {
                setResult(formatBoolWire(item, text));
            }

            @Override
            public void visit(EditableLargeNumberMenuItem item) {
                setResult(formatLargeNumWire(item, text));
            }

            @Override
            public void visit(Rgb32MenuItem item) {
                setResult(formatRgbItemWire(item, text));
            }

            @Override
            public void visit(ScrollChoiceMenuItem item) {
                setResult(formatScrollItemWire(item, text));
            }

            @Override
            public void visit(EditableTextMenuItem item) {
                setResult(formatEditableTextWire(item, text));
            }
        }).orElseThrow();
    }

    private String formatEditableTextWire(EditableTextMenuItem et, String text) {
        if (et.getItemType() == EditItemType.PLAIN_TEXT && text.length() < et.getTextLength()) {
            return text;
        } else if (et.getItemType() == EditItemType.IP_ADDRESS) {
            if (!text.matches("\\d+\\.\\d+\\.\\d+\\.\\d+")) return "0.0.0.0";
            return text;
        } else if (et.getItemType() == EditItemType.TIME_24H || et.getItemType() == EditItemType.TIME_24_HUNDREDS || et.getItemType() == EditItemType.TIME_12H) {
            // time is always sent back to the server in 24 hour format, it is always possible (but optional) to provide hundreds/sec.
            if (!text.matches("\\d+:\\d+:\\d+(.\\d*)*")) return "12:00:00";
            return text;
        } else if (et.getItemType() == EditItemType.GREGORIAN_DATE) {
            if (!text.matches("\\d+/\\d+/\\d+")) return "01/01/2000";
            return text;
        }
        return "";
    }

    private static String formatLargeNumWire(EditableLargeNumberMenuItem ln, String text) {
        BigDecimal val;
        try {
            val = new BigDecimal(text);
        } catch (Exception ex) {
            return "0";
        }
        return val.toString();
    }

    private String formatRgbItemWire(Rgb32MenuItem rgb, String text) {
        return new PortableColor(text).toString();
    }

    private String formatScrollItemWire(ScrollChoiceMenuItem scroll, String text) {
        int val = 0;
        try {
            val = Integer.parseInt(text);
        } catch (Exception ex) {
            // ignored
        }
        return new CurrentScrollPosition(val, text).toString();
    }

    private String formatBoolWire(BooleanMenuItem bi, String text) {
        text = text.toUpperCase();
        if (text.equals("ON") || text.equals("YES") || text.equals("TRUE")) return "1";
        else if (text.equals("OFF") || text.equals("NO") || text.equals("FALSE")) return "0";
        return "0";
    }

    private String formatEnumWire(EnumMenuItem en, String text) {
        return text;
    }

    private int GetActualDecimalDivisor(int divisor) {
        if (divisor < 2) return 1;
        return (divisor > 1000) ? 10000 : (divisor > 100) ? 1000 : (divisor > 10) ? 100 : 10;
    }

    private String formatAnalogWire(AnalogMenuItem an, String text) {
        return text;
    }

    public String formatForDisplay(MenuItem item, Object data) {
        if (item == null || data == null) return "";

        if (item instanceof FloatMenuItem) {
            return formatFloatForDisplay((FloatMenuItem) item, (float) data);
        } else if (item instanceof AnalogMenuItem) {
            return formatAnalogForDisplay((AnalogMenuItem) item, (int) data);
        } else if (item instanceof BooleanMenuItem) {
            return formatBoolForDisplay((BooleanMenuItem) item, (boolean) data);
        } else if (item instanceof EnumMenuItem) {
            return formatEnumForDisplay((EnumMenuItem) item, (int) data);
        } else if (item instanceof EditableLargeNumberMenuItem) {
            return formatLargeNumForDisplay((EditableLargeNumberMenuItem) item, (BigDecimal) data);
        } else if (item instanceof EditableTextMenuItem) {
            return formatTextForDisplay((EditableTextMenuItem) item, (String) data);
        } else if (item instanceof Rgb32MenuItem) {
            return formatRgbItemForDisplay((Rgb32MenuItem) item, (PortableColor) data);
        } else if (item instanceof ScrollChoiceMenuItem) {
            return formatScrollItemForDisplay((ScrollChoiceMenuItem) item, (CurrentScrollPosition) data);
        } else {
            return "";
        }
    }

    private String formatScrollItemForDisplay(ScrollChoiceMenuItem sc, CurrentScrollPosition data) {
        return data.getValue();
    }

    private String formatRgbItemForDisplay(Rgb32MenuItem rgb, PortableColor col) {
        return col.toString();
    }

    private String formatTextForDisplay(EditableTextMenuItem tm, String data) {
        return data;
    }

    private String formatLargeNumForDisplay(EditableLargeNumberMenuItem ln, BigDecimal data) {
        return data.toString();
    }

    private String formatEnumForDisplay(EnumMenuItem en, int data) {
        if (en.getEnumEntries().size() > data) {
            return en.getEnumEntries().get(data);
        }
        return "";
    }

    private String formatBoolForDisplay(BooleanMenuItem bl, boolean val) {
        switch (bl.getNaming()) {
            case ON_OFF:
                return val ? "On" : "Off";
            case YES_NO:
                return val ? "Yes" : "No";
            case TRUE_FALSE:
            default:
                return val ? "True" : "False";
        }
    }

    private String formatAnalogForDisplay(AnalogMenuItem an, int val) {
        int calcVal = val + an.getOffset();
        int divisor = an.getDivisor();

        if (divisor < 2) {
            return Integer.toString(calcVal) + bundleIfPossible(an.getUnitName());
        } else {
            int whole = calcVal / divisor;
            int fractMax = GetActualDecimalDivisor(an.getDivisor());
            int fraction = Math.abs((calcVal % divisor)) * (fractMax / divisor);

            return String.format("%d.%0" + calculateRequiredDigits(divisor) + "d%s", whole, fraction,
                    localHandler.getFromLocaleOrUseSource(an.getUnitName()));
        }
    }

    private int calculateRequiredDigits(int divisor) {
        return (divisor <= 10) ? 1 : (divisor <= 100) ? 2 : (divisor <= 1000) ? 3 : 4;
    }

    private String formatFloatForDisplay(FloatMenuItem fl, float val) {
        return String.format("%." + fl.getNumDecimalPlaces() + "f", val);
    }

    public String getItemName(MenuItem item) {
        return localHandler.getFromLocaleWithDefault(item.getName(), item.getName());
    }
}

File: web-designer/tcmenugen/src/domain/MenuItemFormatter.ts
import {
    AnalogMenuItem,
    BooleanMenuItem,
    BooleanNaming,
    EditableLargeNumberMenuItem,
    EditableTextMenuItem,
    EnumMenuItem,
    FloatMenuItem, ListMenuItem,
    MenuItem,
    Rgb32MenuItem,
    ScrollChoiceMenuItem, TcMenuItemError, TextEditMode
} from "./MenuItem";

function isTrue(currentValue: string): boolean {
    return currentValue.charAt(0) === "Y" || currentValue.charAt(0) === "1" || currentValue.charAt(0) === "T";
}

export function formatStringToWire(item: MenuItem<any>, currentValue: string): string {
    if(item instanceof AnalogMenuItem) {
        let num = parseFloat(currentValue);
        num = Math.round((num * item.getDivisor()) - item.getOffset());
        if(num < 0 || num > item.getMaxValue()) throw new TcMenuItemError(`Number ${num} outside of 0..${item.getMaxValue()}`);
        return num.toFixed(0);
    }
    else if(item instanceof EnumMenuItem) {
        let num = parseInt(currentValue);
        if(num < 0 || num >= item.getItemList().length) throw new TcMenuItemError(`Enum ${num} outside allowable range`);
        return item.getCurrentValue().toFixed(0);
    }
    else if(item instanceof BooleanMenuItem) return isTrue(currentValue) ? "1" : "0";
    else if(item instanceof EditableLargeNumberMenuItem) {
        let flt = parseFloat(currentValue);
        if(flt < 0 && !item.isNegativeAllowed()) throw new TcMenuItemError("Negative value not allowed");
        return flt.toFixed(item.getDecimalPlaces());
    }
    else if(item instanceof EditableTextMenuItem) {
        return formatEditableTextWire(item, currentValue);
    }
    else if(item instanceof Rgb32MenuItem) {
        if(!currentValue.match("#[0-9A-Fa-f]*")) throw new TcMenuItemError("Not in HTML color format");
        return currentValue;
    }
    else if(item instanceof ScrollChoiceMenuItem)  {
        let num = parseInt(currentValue);
        return num.toFixed(0) + "-";
    }
    else throw new TcMenuItemError("Unknown type of item for text conversion");
}

export function formatForDisplay(item: MenuItem<any>): string {
    if(item instanceof FloatMenuItem) return item.getCurrentValue().toFixed(item.getDecimalPlaces());
    else if(item instanceof AnalogMenuItem) return formatAnalogItem(item);
    else if(item instanceof EnumMenuItem) return item.getItemList()[item.getCurrentValue()];
    else if(item instanceof BooleanMenuItem) return formatBooleanItem(item);
    else if(item instanceof EditableTextMenuItem) return item.getCurrentValue();
    else if(item instanceof EditableLargeNumberMenuItem) return item.getCurrentValue().toFixed(item.getDecimalPlaces());
    else if(item instanceof Rgb32MenuItem) return item.getCurrentValue();
    else if(item instanceof ScrollChoiceMenuItem) return item.getCurrentValue().currentValue;
    else if(item instanceof ListMenuItem) return Object.values(item.getCurrentValue()).join(", ");
    else return "";
}

function formatBooleanItem(item: BooleanMenuItem) {
    let curr = item.getCurrentValue();
    switch(item.getNaming()) {
        case BooleanNaming.YES_NO: return curr ? "YES" : "NO";
        case BooleanNaming.ON_OFF: return curr ? "ON" : "OFF";
        default: return curr ? "TRUE" : "FALSE";
    }
}

function getActualDecimalDivisor(divisor: number): number {
    if (divisor < 2) return 1;
    return (divisor > 1000) ? 10000 : (divisor > 100) ? 1000 : (divisor > 10) ? 100 : 10;
}

function calculateRequiredDigits(divisor: number) {
    return (divisor <= 10) ? 1 : (divisor <= 100) ? 2 : (divisor <= 1000) ? 3 : 4;
}

function formatAnalogItem(an: AnalogMenuItem) {
    return formatAnalogValue(an, an.getCurrentValue() + an.getOffset());
}

export function formatAnalogValue(an: AnalogMenuItem, val: number) {
    let calcVal = val;
    let divisor = an.getDivisor();

    if (divisor < 2)  {
        return calcVal.toFixed(0) + an.getUnitName();
    } else {
        let whole = Math.floor(calcVal / divisor);
        let fractMax = getActualDecimalDivisor(an.getDivisor());
        let fraction = (Math.abs((calcVal % divisor)) * (fractMax / divisor)).toFixed(0);
        return whole.toFixed(0) + "." + fraction.padStart(calculateRequiredDigits(divisor), "0") + an.getUnitName();
    }
}

function formatEditableTextWire(et: EditableTextMenuItem, val: string)  {
    if(et.getEditMode() === TextEditMode.PLAIN_TEXT) {
        if(val.length > et.getTextLength()) throw new TcMenuItemError("Text too long");
        return val;
    }
    else if(et.getEditMode() === TextEditMode.IP_ADDRESS)
    {
        if (!val.match("\\d+\\.\\d+\\.\\d+\\.\\d+")) throw new TcMenuItemError("Not an IPV4 address");
        return val;
    }
    else if(et.getEditMode() === TextEditMode.TIME_24H || et.getEditMode() === TextEditMode.TIME_24_HUNDREDS || et.getEditMode() === TextEditMode.TIME_12H)
    {
        // time is always sent back to the server in 24 hour format, it is always possible (but optional) to provide hundreds/sec.
        if (!val.match("\\d+:\\d+:\\d+(.\\d*)*")) throw new TcMenuItemError("Not in the correct time format");
        return val;
    }
    else if (et.getEditMode() === TextEditMode.GREGORIAN_DATE)
    {
        if (!val.match("\\d+/\\d+/\\d+")) throw new TcMenuItemError("Not a date");
        return val;
    }
    return "";
}

export class AnalogStarter {
    private readonly _name: string;
    private readonly _offset: number;
    private readonly _divisor: number;
    private readonly _maxValue: number;
    private readonly _unitName: string;

    constructor(name: string, offset: number, divisor: number, maxValue: number, unitName: string) {
        this._name = name;
        this._offset = offset;
        this._divisor = divisor;
        this._maxValue = maxValue;
        this._unitName = unitName;
    }

    get name(): string {
        return this._name;
    }

    get offset(): number {
        return this._offset;
    }

    get divisor(): number {
        return this._divisor;
    }

    get maxValue(): number {
        return this._maxValue;
    }

    get unitName(): string {
        return this._unitName;
    }
}



export let TC_ANALOG_STARTERS: AnalogStarter[] = [
    new AnalogStarter("Integer 0->255A", 0, 1, 255, "A"),
    new AnalogStarter("Integer 1->500m", 1, 1, 499, "m"),
    new AnalogStarter("Percentage 0->100%", 0, 1, 100, "%"),
    new AnalogStarter("Integer -100->100", -100, 1, 100, ""),
    new AnalogStarter("Volume -90.0->37.5dB", -180, 2, 255, "dB"),
    new AnalogStarter("Tenths 0.0->100.0", 0, 10, 1000, ""),
    new AnalogStarter("Hundredths -1.00->1.00V", -100, 100, 200, "V"),
    new AnalogStarter("Negative Tenths -10.0->0.0V", -100, 100, 100, "V")
];